### PR TITLE
Native `std::atomic`, attempt 1.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich ,

On my machine, this compiles file with `clang++`, and fails to link with `g++.

```
dima@dima-i9: ~/github/dkorolev/Current/Blocks/Persistence(native_atomics) $ clang++ --version
Ubuntu clang version 3.5.0-4ubuntu2~trusty2 (tags/RELEASE_350/final) (based on LLVM 3.5.0)
Target: x86_64-pc-linux-gnu
Thread model: posix
dima@dima-i9: ~/github/dkorolev/Current/Blocks/Persistence(native_atomics) $ g++ --version
g++ (Ubuntu 4.8.4-2ubuntu1~14.04.1) 4.8.4
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

dima@dima-i9: ~/github/dkorolev/Current/Blocks/Persistence(native_atomics) $ export CPLUSPLUS=clang++
dima@dima-i9: ~/github/dkorolev/Current/Blocks/Persistence(native_atomics) $ make clean test
find . -type d -name .current | xargs rm -rf
find . -type l -name '.current_*' | xargs rm -f
find . -type l -name '.current_*.*' | xargs rm -f
find . -name core | xargs rm -f
mkdir -p .current
clang++ -std=c++11 -Wall -W -latomic -o .current/test test.cc -pthread
.current/test --current_runtime_arch=Linux
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from PersistenceLayer
[ RUN      ] PersistenceLayer.Memory
[       OK ] PersistenceLayer.Memory (0 ms)
[ RUN      ] PersistenceLayer.File
[       OK ] PersistenceLayer.File (2 ms)
[ RUN      ] PersistenceLayer.Exceptions
[       OK ] PersistenceLayer.Exceptions (2 ms)
[----------] 3 tests from PersistenceLayer (4 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (4 ms total)
[  PASSED  ] 3 tests.
dima@dima-i9: ~/github/dkorolev/Current/Blocks/Persistence(native_atomics) $ export CPLUSPLUS=g++
dima@dima-i9: ~/github/dkorolev/Current/Blocks/Persistence(native_atomics) $ make clean test
find . -type d -name .current | xargs rm -rf
find . -type l -name '.current_*' | xargs rm -f
find . -type l -name '.current_*.*' | xargs rm -f
find . -name core | xargs rm -f
mkdir -p .current
g++ -std=c++11 -Wall -W -latomic -o .current/test test.cc -pthread
/tmp/ccYOv2x0.o: In function `std::atomic<current::persistence::impl::FilePersister<persistence_test::CURRENT_STRUCT_IMPL_StorableString<current::reflection::DeclareFields> >::end_t>::load(std::memory_order) const':
test.cc:(.text._ZNKSt6atomicIN7current11persistence4impl13FilePersisterIN16persistence_test34CURRENT_STRUCT_IMPL_StorableStringINS0_10reflection13DeclareFieldsEEEE5end_tEE4loadESt12memory_order[_ZNKSt6atomicIN7current11persistence4impl13FilePersisterIN16persistence_test34CURRENT_STRUCT_IMPL_StorableStringINS0_10reflection13DeclareFieldsEEEE5end_tEE4loadESt12memory_order]+0x28): undefined reference to `__atomic_load_16'
/tmp/ccYOv2x0.o: In function `std::atomic<current::persistence::impl::FilePersister<persistence_test::CURRENT_STRUCT_IMPL_StorableString<current::reflection::DeclareFields> >::end_t>::store(current::persistence::impl::FilePersister<persistence_test::CURRENT_STRUCT_IMPL_StorableString<current::reflection::DeclareFields> >::end_t, std::memory_order)':
test.cc:(.text._ZNSt6atomicIN7current11persistence4impl13FilePersisterIN16persistence_test34CURRENT_STRUCT_IMPL_StorableStringINS0_10reflection13DeclareFieldsEEEE5end_tEE5storeESA_St12memory_order[_ZNSt6atomicIN7current11persistence4impl13FilePersisterIN16persistence_test34CURRENT_STRUCT_IMPL_StorableStringINS0_10reflection13DeclareFieldsEEEE5end_tEE5storeESA_St12memory_order]+0x44): undefined reference to `__atomic_store_16'
collect2: error: ld returned 1 exit status
make: *** [.current/test] Error 1
```

Keeping in mind we also need to support Windows, I suspect the easiest solution from here is conditional compilation.

Suggestions welcome.

Thanks!
Dima

### CODE REVIEW

**Comment by dkorolev** [comment]
@mzhurovich , for `g++`, FWIW, `-mcx16` has no effect, both on its own and in conjunction with `-latomic`.

**Comment by dkorolev** [comment]
Status update: We can compile it (and just `-latomic` is enough), but `g++` version results in a segfault in `atomic::load`.

Falling back to the poor man's solution.

File: Blocks/Persistence/file.h [hunk 1]
```diff
@@ -92,35 +94,49 @@ class IteratorOverFileOfPersistedEntries {
...
+          // Read through all the lines.
+          // Let `IteratorOverFileOfPersistedEntries` maintain its own `next_`, which later becomes `this->end`.
+          ;
         }
-        return cit.Next();
+        const auto& end = cit.Next();
+        return end_t{end.index, end.us};
       } else {
-        return idxts_t();
+        return end_t{0ull, std::chrono::microseconds(0)};
```
Comment (mzhurovich) [inline Blocks/Persistence/file.h]: We could add custom constructor, but brace-init may be the safest choice :)

Comment (dkorolev) [inline Blocks/Persistence/file.h]: Rule of zero: http://en.cppreference.com/w/cpp/language/rule_of_three

I figured for `atomic`-s it's more important than a proper constructor.

**Pull request merged**